### PR TITLE
[OTX/XAI] support xai baseline for OTX

### PR DESCRIPTION
**Pull request by dongkwan-kim01**
This is the same as #72 but the target branch changed to `otx`.

---

**Summary**
Support explain mode to MPA classification

**This PR includes**
Add ClsExplainer class to mpa/cls/explainer.py
Add explain stage

**Test Codes**
MPA standalone code will be deprecated
Can test with OTX CLI(please refer OTX cli support, [link](https://github.com/openvinotoolkit/training_extensions/pull/1342))

### CODE REVIEW

**Comment by dongkwan-kim01** [comment]
@negvet @wonjuleee @harimkang @sungmanc Hello, could you re-review this PR? Test code passed with OTX cli.

File: mpa/cls/explainer.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        if cfg.load_from is not None:
+            logger.info('load checkpoint from ' + cfg.load_from)
+            _ = load_checkpoint(model, cfg.load_from, map_location='cpu')
+
+        model.eval()
+        model = MMDataParallel(model, device_ids=[0])
+
+        # InferenceProgressCallback (Time Monitor enable into Infer task)
+        ClsStage.set_inference_progress_callback(model, cfg)
+        with self.explainer_hook(model.module.backbone) as shook:
```
Comment (sungmanc) [inline mpa/cls/explainer.py]: Minor comment: how about changing the variable name `shook`? because I think `shook` can't deliver the meaning of the variable.
Like, `backward_explainer_hook` ?

File: mpa/modules/xai/builder.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+def build_explainer(model, cfg):
```
Comment (sungmanc) [inline mpa/modules/xai/builder.py:1]: It seems that this function is not called and has no meaning, if I am right, how about remove?

File: recipes/cls_explain.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+workflow_hooks:
+    - type: WFProfileHook
+      name: profile_hook
+
+stages:
+  - name: cls-explanation
+    type: ClsExplainer
+    task: classification
+    mode: explain
+    config: {{ fileDirname }}/stages/classification/class_incr.yaml
```
Comment (sungmanc) [inline recipes/cls_explain.yaml]: Is it okay to use the configs from `class_incr.yaml` ? If `explainer` needs special configs, then how about using independent config file?

File: mpa/cls/explainer.py (hunk 1 above)
Comment (dongkwan-kim01) [reply mpa/cls/explainer.py]: Fixed. Thank you for good suggestion

File: mpa/modules/hooks/auxiliary_hooks.py [hunk 4]
```diff
@@ -13,23 +13,48 @@
...
+from abc import ABC, abstractmethod
 from typing import Union
 
 import torch
 
 
-class EigenCamHook:
-    def __init__(self, module: torch.nn.Module) -> None:
+class BaseAuxiliaryHook(ABC):
+    def __init__(self, module: torch.nn.Module, _fpn_idx: int = 0) -> None:
```
Comment (goodsong81) [inline mpa/modules/hooks/auxiliary_hooks.py]: The functionality of this class is not Auxiliary anymore.
How about "RecordingForwardHook", "ModuleRecordingHook", etc?
BTW, don't use _xxxx name for public parameters. "_fpn_idx"???

File: mpa/modules/hooks/auxiliary_hooks.py [hunk 5]
```diff
@@ -106,91 +102,49 @@ def func(feature_map: Union[torch.Tensor, list[torch.Tensor]], _fpn_idx: int = 0
             feature_map = feature_map[_fpn_idx]
```
Comment (sungmanc) [inline mpa/modules/hooks/auxiliary_hooks.py]: How about add `assert` or `raise` if `fpn_idx` is out of scope?
i.e) When the FPN has [0,1,2] indices, if user input the variable `fpn_idx` as 3, we need to description.

File: mpa/modules/hooks/auxiliary_hooks.py [hunk 6]
```diff
@@ -46,57 +71,28 @@ def func(x: torch.Tensor) -> torch.Tensor:
...
     Example::
-        with SaliencyMapHook(model.module.backbone) as hook:      
+        with ActivationMapHook(model.module.backbone) as hook:
             with torch.no_grad():
                 result = model(return_loss=False, **data)
             print(hook.records)
     Args:
         module (torch.nn.Module): The PyTorch module to be registered in forward pass
-        _fpn_idx (int, optional): The layer index to be processed if the model is a FPN. 
+        _fpn_idx (int, optional): The layer index to be processed if the model is a FPN.
```
Comment (goodsong81) [inline mpa/modules/hooks/auxiliary_hooks.py]: Rename this.

File: mpa/modules/hooks/auxiliary_hooks.py [hunk 7]
```diff
@@ -46,57 +71,28 @@ def func(x: torch.Tensor) -> torch.Tensor:
...
     @staticmethod
     def func(feature_map: Union[torch.Tensor, list[torch.Tensor]], _fpn_idx: int = 0) -> torch.Tensor:
         """Generate the saliency map by average feature maps then normalizing to (0, 255).
 
         Args:
-            feature_map (Union[torch.Tensor, list[torch.Tensor]]): feature maps from backbone or list of feature maps 
+            feature_map (Union[torch.Tensor, list[torch.Tensor]]): feature maps from backbone or list of feature maps
                                                                     from FPN.
-            _fpn_idx (int, optional): The layer index to be processed if the model is a FPN. 
+            _fpn_idx (int, optional): The layer index to be processed if the model is a FPN.
```
Comment (goodsong81) [inline mpa/modules/hooks/auxiliary_hooks.py]: Same here

**Review by goodsong81 — CHANGES_REQUESTED** [review]
Thank you for the update.
Could you please refine some class/variable names? Others looks good to me.

File: mpa/modules/xai/builder.py (hunk 2 above)
Comment (dongkwan-kim01) [reply mpa/modules/xai/builder.py:1]: This remained for the initial design suggestion but can be removed, and added when future implemented in the future.

File: mpa/modules/hooks/auxiliary_hooks.py (hunk 5 above)
Comment (goodsong81) [reply mpa/modules/hooks/auxiliary_hooks.py]: It will raise exception here anyway. Do we need to do this unless we are taking care of the exception?

Comment (sungmanc) [reply mpa/modules/hooks/auxiliary_hooks.py]: No, we don't need it. Just want to give a kind description. i.e. by checking the error message, users easily find the real FPN range.

**Comment by dongkwan-kim01** [comment]
All comments applied, @goodsong81 @sungmanc. Thank you for suggestions.

**Comment by sungmanc** [comment]
LGTM

**Review by sungmanc — APPROVED** [review]
LGTM

**Comment by sungmanc** [comment]
BTW, how about control the frequency of push? push with every commit is needed?

**Comment by dongkwan-kim01** [comment]
> BTW, how about control the frequency of push? push with every commit is needed?

Yeah, sorry for the frequent alarm. I'll try to reduce the frequency.

**Review by negvet — APPROVED** [review]
LGTM

File: mpa/modules/hooks/auxiliary_hooks.py (hunk 4 above)
Comment (dongkwan-kim01) [reply mpa/modules/hooks/auxiliary_hooks.py]: Yes, as you said recording hook is more appropriate. The hook name changed and `_` removed from the parameter name.

File: models/classification/ote_efficientnet_b0_multilabel_al.yaml [hunk 8]
```diff
@@ -2,11 +2,9 @@ _base_: ./ote_efficientnet_b0_multilabel.yaml
 
 model:
     head:
-        type: CustomMultiLabelNonLinearClsHead
+        type: CustomMultiLabelLinearClsHead
```
Comment (JihwanEom) [inline models/classification/ote_efficientnet_b0_multilabel_al.yaml:5]: Is it okay @kprokofi @sovrasov?

File: models/classification/ote_efficientnet_v2_s_multilabel_al.yaml [hunk 9]
```diff
@@ -2,11 +2,9 @@ _base_: ./ote_efficientnet_v2_s_multilabel.yaml
 
 model:
     head:
-        type: CustomMultiLabelNonLinearClsHead
+        type: CustomMultiLabelLinearClsHead
```
Comment (JihwanEom) [inline models/classification/ote_efficientnet_v2_s_multilabel_al.yaml:5]: Same with here

File: recipes/stages/_base_/data/pipelines/semisl_pipeline.py [hunk 10]
```diff
@@ -23,7 +23,7 @@
 
 train_pipeline_strong = [
     dict(type="Resize", size=__resize_target_size),
-    dict(type="RandAugment", n=2, m=10),
+    dict(type="MPARandAugment", n=2, m=10),
```
Comment (JihwanEom) [inline recipes/stages/_base_/data/pipelines/semisl_pipeline.py:26]: How about `OTXRandAugment`? It's very very minor suggestion :)

File: recipes/stages/instance-segmentation/imbalance.py [hunk 11]
```diff
@@ -6,10 +6,6 @@
 
 task = 'instance-segmentation'
 
-data = dict(
```
Comment (JihwanEom) [inline recipes/stages/instance-segmentation/imbalance.py:9]: Do you have specific reason to remove this code? That can be needed to support semi-supervised learning.

File: models/classification/ote_efficientnet_b0_multilabel_al.yaml (hunk 8 above)
Comment (sovrasov) [reply models/classification/ote_efficientnet_b0_multilabel_al.yaml:5]: Yep, the head should be linear for EfficientNets as it is in [develop](https://github.com/openvinotoolkit/model_preparation_algorithm/blob/develop/models/classification/ote_efficientnet_b0_multilabel_al.yaml) branch now. I have no idea why it was changes in otx, but this PR fixes the issue.

Comment (dongkwan-kim01) [reply models/classification/ote_efficientnet_b0_multilabel_al.yaml:5]: I don't know why this change is here, maybe the push from [8c4c328](https://github.com/openvinotoolkit/model_preparation_algorithm/commit/8c4c32840016c254e5e23addac0063633559e1d7)?

Comment (JihwanEom) [reply models/classification/ote_efficientnet_b0_multilabel_al.yaml:5]: Anyway, this PR fixed multi-label bugs :) LGTM

File: mpa/cls/explainer.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+from mmcls.models import build_classifier
+
+from mpa.registry import STAGES
+from mpa.cls.stage import ClsStage
+from mpa.modules.hooks.recording_forward_hooks import ActivationMapHook, EigenCamHook
+from mpa.utils.logger import get_logger
+logger = get_logger()
+EXPLAINER_HOOK_SELECTOR = {
+    'eigencam': EigenCamHook,
+    'activationmap': ActivationMapHook}
```
Comment (goodsong81) [inline mpa/cls/explainer.py]: Is this right format? You may want to run local pre-commit test if CI is not working well.

File: mpa/cls/explainer.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        outputs = self._explain(cfg)
+        return dict(outputs=outputs)
+
+    def _explain(self, cfg):
+        if cfg.get('task_adapt', False) and not hasattr(self, 'eval'):
+            dataset_cfg = cfg.data.train.copy()
+            dataset_cfg.pipeline = cfg.data.test.pipeline
+            self.dataset = build_dataset(dataset_cfg)
+        else:
+            self.dataset = build_dataset(cfg.data.test)
```
Comment (goodsong81) [inline mpa/cls/explainer.py]: Could you refactor this block to a method with proper name?

File: mpa/cls/explainer.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        data_loader = build_dataloader(
+            self.dataset,
+            samples_per_gpu=cfg.data.samples_per_gpu,
+            workers_per_gpu=cfg.data.workers_per_gpu,
+            dist=False,
+            shuffle=False,
+            round_up=False,
+            persistent_workers=False)
+
+        print(cfg.explainer)
```
Comment (goodsong81) [inline mpa/cls/explainer.py]: Remove print.
Or use logger explicitly, if needed.

File: mpa/cls/explainer.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+        # build the model and load checkpoint
+        model = build_classifier(cfg.model)
+        self.extract_prob = hasattr(model, 'extract_prob')
+        fp16_cfg = cfg.get('fp16', None)
+        if fp16_cfg is not None:
+            wrap_fp16_model(model)
+        if cfg.load_from is not None:
+            logger.info('load checkpoint from ' + cfg.load_from)
```
Comment (goodsong81) [inline mpa/cls/explainer.py]: -> "Load checkpoint from "

File: mpa/cls/explainer.py (hunk 12 above)
Comment (goodsong81) [reply mpa/cls/explainer.py]: Found inconsistency across this PR: 'str' or "str"?

File: mpa/det/trainer.py [hunk 16]
```diff
@@ -153,11 +154,12 @@ def train_worker(gpu, target_classes, datasets, cfg, distributed=False,
...
-            if getattr(cfg.model.bbox_head.anchor_generator, 'reclustering_anchors', False):
-                train_cfg = Stage.get_train_data_cfg(cfg)
-                train_dataset = train_cfg.get('ote_dataset', None)
-                cfg, model = cluster_anchors(cfg, train_dataset, model)
+        # TODO[JAEGUK]: Temporal Disable cluster_anchors for SSD model
+        # if hasattr(cfg.model, 'bbox_head') and hasattr(cfg.model.bbox_head, 'anchor_generator'):
+        #     if getattr(cfg.model.bbox_head.anchor_generator, 'reclustering_anchors', False):
+        #         train_cfg = Stage.get_train_data_cfg(cfg)
+        #         train_dataset = train_cfg.get('otx_dataset', None)
+        #         cfg, model = cluster_anchors(cfg, train_dataset, model)
```
Comment (goodsong81) [inline mpa/det/trainer.py:169]: Is this relevant change?
@jaegukhyun Could you help on this?

File: requirements.txt [hunk 17]
```diff
@@ -7,5 +7,5 @@ networkx>=2.7.1,<=2.8.0
 defusedxml>=0.7.1
 tensorboard>=2.4.1
 scikit-image>=0.17.2
-torch>=1.7.1,<=1.9.0
-torchvision>=0.8.1,<=0.10.0
+torch>=1.8.1, <=1.9.1
+torchvision>=0.8.1,<=0.10.1
```
Comment (goodsong81) [inline requirements.txt:11]: Why did you?

File: setup.py [hunk 18]
```diff
@@ -2,33 +2,36 @@
...
     long_description = fh.read()
 
 
-def get_requirements(filename='requirements.txt'):
-    here = osp.dirname(osp.realpath(__file__))
-    with open(osp.join(here, filename), 'r') as f:
-        requires = [line.replace('\n', '') for line in f.readlines()]
+def get_requirements(filename="requirements.txt"):
+    here = os.path.dirname(os.path.realpath(__file__))
+    with open(os.path.join(here, filename), "r") as f:
```
Comment (goodsong81) [inline setup.py:15]: Specific reason for osp -> os.path?

**Review by goodsong81 — CHANGES_REQUESTED** [review]

File: mpa/cls/explainer.py (hunk 12 above)
Comment (dongkwan-kim01) [reply mpa/cls/explainer.py]: Yes, all changed to 'str', following other files and the `EXPLAINER_HOOK_SELECTOR` one does not occur error in pre-commit but modified to split lines for readability.

File: mpa/cls/explainer.py (hunk 13 above)
Comment (dongkwan-kim01) [reply mpa/cls/explainer.py]: Fixed to `explain_dataset` and removed if - else part

File: mpa/cls/explainer.py (hunk 14 above)
Comment (dongkwan-kim01) [reply mpa/cls/explainer.py]: Fixed

File: mpa/cls/explainer.py (hunk 15 above)
Comment (dongkwan-kim01) [reply mpa/cls/explainer.py]: Fixed

File: mpa/det/trainer.py (hunk 16 above)
Comment (dongkwan-kim01) [reply mpa/det/trainer.py:169]: This one is maybe from push to `otx`, don't know why but will close this PR and this PR moved to #87

File: requirements.txt (hunk 17 above)
Comment (dongkwan-kim01) [reply requirements.txt:11]: Same reason as above

File: setup.py (hunk 18 above)
Comment (dongkwan-kim01) [reply setup.py:15]: Same reason as above

**Comment by dongkwan-kim01** [comment]
This closed by moving to #87
